src/Columns/ColumnArray.cpp

```diff
@@ -524,23 +524,21 @@ void ColumnArray::insertRangeFrom(const IColumn & src, size_t start, size_t leng
     size_t nested_offset = src_concrete.offsetAt(start);
     size_t nested_length = src_concrete.getOffsets()[start + length - 1] - nested_offset;
 
+    Offsets & cur_offsets = getOffsets();
+    cur_offsets.reserve(cur_offsets.size() + length);
```

Comment:
you achieved ex safety by pre-allocating memory before starting to modify smth, right? but why it is necessary to replace `resize` with `reserve` and start to use `push_back`?

Reply:
>you achieved ex safety by pre-allocating memory before starting to modify smth, right?

Yep.

> but why it is necessary to replace resize with reserve and start to use push_back?

Because otherwise you will get incorrect number of elements after exception from insertRangeFrom, that is default-initialized.

Reply:
but what if we just "book" this memory with `reserve` and left everything else untouched?

Reply:
Then it will not be seen by anyone who iterates through it.

Consider the following code:

```c
PODArray<UInt8> chars;
PODArray<size_t> offsets;

chars.resize(new_size);
// fill chars

offsets.resize(new_offset_size); // throws
// so now chars.size() > offsets.size()
```

While this code is more safe:

```c
PODArray<UInt8> chars;
PODArray<size_t> offsets;

offsets.reserve(new_offset_size); // reserve does not update size, we are safe
chars.resize(new_size); // if this throws, size will not be updated, we are safe
// But sure while you will fill `chars` you may get some exception, and in this case offsets.size() will not match with chars.size(), but this is more unlikely than throw from resize()
```

Reply:
I mean

``` cpp
PODArray<UInt8> chars;
PODArray<size_t> offsets;

offsets.reserve(new_offset_size); // size() == 0 and now offsets.resize() won't throw
chars.resize(new_size); // doesn't matter if throws
offsets.resize(new_offset_size);
for (...)
    offsets[i] = ...;
```

Reply:
Oh, I see, sounds reasonable, nice catch! Fixed.